### PR TITLE
extend Version struct with version field

### LINKED ISSUE #286: pub struct Version is missing version field
```bash
 ~#@❯ hyprctl version -j
{
    "branch": "",
    "commit": "0fb9a04526b06adfb3fd16b64e13b7d110ae7855",
    "version": "0.44.0",
    "dirty": false,
    "commit_message": "",
    "commit_date": "2024-11-04",
    "tag": "",
    "commits": "5409",
    "buildAquamarine": "0.4.3",
    "flags": ["debug"]
}
```
and the struct only contains:
```rust
pub struct Version {
    /// The git branch Hyprland was built on
    pub branch: String,
    /// The git commit Hyprland was built on
    pub commit: String,
    /// This is true if there were unstaged changed when Hyprland was built
    pub dirty: bool,
    /// The git commit message
    pub commit_message: String,
    /// The git commit date
    pub commit_date: String,
    /// The git tag hyprland was built on
    pub tag: String,
    /// The amount of commits to Hyprland at buildtime
    pub commits: String,
    /// Aquamarine version
    #[serde(rename = "buildAquamarine")]
    pub build_aquamarine: String,
    /// The flags that Hyprland was built with
    pub flags: Vec<String>,
}
impl_on!(Version);
```

would make sense to extend the struct, as for hyprland-git versions the tag is not filled and there is another way needed to get the version.

### PR DESCRIPTION
**Pull request by simon-bueler**
resolves #286

### CODE REVIEW

**Comment by H3rmt** [comment]
It might be beneficial to make this struct as backwards compatible as possible.

I use this to check if the current Hyprland Instance is newer than a certain version to prevent users from getting weird serde errors.

I don't exactly know when the version field was added but this would mean hyprswitch wouldn't be able to check the version of hyprswitch lower than the version the "version" field was added to the Version output.

Then users would again get a weird serde error instead of a warning explaining that they are running version 0.41 and need at least 0.42

**Comment by H3rmt** [comment]
v0.44.1 (newest from aur has no version field)

**Comment by H3rmt** [comment]
added with https://github.com/hyprwm/Hyprland/commit/7564b26b7d386d248eaa47c1a481c09eefd8e3ca

**Comment by H3rmt** [comment]
image

**Comment by H3rmt** [comment]
using 
```rs
#[serde(default)]
/// The Hyprland version
pub version: String,
```
returns "" as the Version string if the Hyprland version is to new

**Review by yavko — APPROVED** [review]

**Comment by yavko** [comment]
> using 
> ```rs
> #[serde(default)]
> /// The Hyprland version
> pub version: String,
> ```
> returns "" as the Version string if the Hyprland version is to new:
> 
> ```
> TRACE Hyprland Version { branch: "", commit: "4520b30d498daca8079365bdb909a8dea38e8d55", version: "", dirty: false, commit_message: "version: bump to 0.44.1", commit_date: "Wed Oct 9 12:54:39 2024", tag: "v0.44.1", commits: "5309", build_aquamarine: "0.4.3", flags: [] }
> ```

Perhaps that's a good idea

**Comment by simon-bueler** [comment]
@H3rmt: as for hyprswitch, if we create a new version of hyprland-rs (probably 0.4.0-beta.2) you can rewrite the check to fit both variants and add the new hyprland-rs to that version (probably hyprswitch 3.0.2 or so)

**Comment by H3rmt** [comment]
I made the version check optional, so hyprswitch continues working if an outdated or old version is found, but it shows a warning if the Tag wasn't available.

This is needed to allow running the version check for <0.44.1 on git version without the tag

**Comment by H3rmt** [comment]
But having the version field set to an empty string (if none was reported) is definitely needed as it would break getting the version even on the current aur release.

**Review by yavko — CHANGES_REQUESTED** [review]
I'd like it not to break fairly recent versions, so i think setting a default would be nice.

**Comment by simon-bueler** [comment]
Sorry for being so incapable but I actually don't know how to add that in rust. Would you be so kind to help me there?

**Comment by H3rmt** [comment]
Do you see the change request on GitHub?
It contains the rust code needed to populate the version attribute with an empty string

**Comment by H3rmt** [comment]
```rs
#[serde(default)]
/// The Hyprland version
pub version: String, 
```

That's the rust code needed to add the version as an "optional" value.

**Comment by yavko** [comment]
Actually, why not make it an option instead, an empty string is less meaningful.

File: src/data/regular.rs [hunk 1]
```diff
@@ -428,6 +428,8 @@ pub struct Version {
     pub branch: String,
     /// The git commit Hyprland was built on
     pub commit: String,
+    /// The Hyprland version
```
[the remark this answers is not in the document]
Comment (simon-bueler) [reply src/data/regular.rs:432]: Its in the second commit

**Review by yavko — CHANGES_REQUESTED** [review]
Try making it into an option instead, it would make more sense that way.

**Review by yavko — APPROVED** [review]
thx

**Comment by H3rmt** [comment]
@yavko they removed nearly every value from the version command (at least they return "" instead of removing the fields)

**Comment by yavko** [comment]
> @yavko they removed nearly every value from the version command (at least they return "" instead of removing the fields)

Bruh, r u sure its not a bug

**Pull request merged**